### ISSUE TITLE
XGBoost4j ranking: Grouping data information may be matched wrong

### ISSUE THREAD

**Issue by lblaese** [dmlc/xgboost#2344]
I want to use XGBoost4j on Spark to rank a list of documents for a given qurey. When looking at the code, I saw, that you have to provide grouping information (see https://github.com/dmlc/xgboost/blob/master/doc/input_format.md#group-input-format) for that.

Let's say you have a number of documents that you want to rank and your query is something like 'give me a ranking of all documents of a certain language'. Then you would group your documents (the training data) by language and train the model. Is that how it is supposed to work?

The grouping data gives, according to its doc, only information on how many training points (let's say documents) are in which group, it does not give a document-wise mapping (it does not say which document is in which group). 
Because the RDD is shuffled (e.g. XGBoost.scala l 107) and therefore the order of the trainingdata, even within a partition, is randomized, this information is essentially worthless. 

I don't really see how the groupingData is used, apart from being set on the DMatrix in XGBoost.scala, so I am not sure if that really is a bug. It would be nice, if anyone would comment on the use of grouping information, how to interpret it and how it is actually used in the code.

Edit: I just saw that the groupData gets used in rank_metric.cc to evaluate e.g. EvalAUC. Is this the only time its needed?

**Comment by CodingCat** [comment]
would you mind titling with `[jvm-packages]` when you refer to a jvm problem?

**Comment by CodingCat** [comment]
see discussions in https://github.com/dmlc/xgboost/issues/2223
